File: sdk/swift/Sources/UCRSDK/ServiceCredential.swift
public let credentialIDMetadataKey = "ucr-service-credential-id-bin"
public let credentialSecretMetadataKey = "ucr-service-credential-secret-bin"

public struct BinaryMetadataEntry: Sendable {
    public let key: String
    public let value: [UInt8]

    public init(key: String, value: [UInt8]) {
        self.key = key
        self.value = value
    }
}

public struct ServiceCredential: Sendable, CustomStringConvertible {
    private let credentialID: [UInt8]
    private let secret: [UInt8]

    public init(credentialID: [UInt8], secret: [UInt8]) {
        self.credentialID = credentialID
        self.secret = secret
    }

    public func binaryMetadata() -> [BinaryMetadataEntry] {
        [
            BinaryMetadataEntry(key: credentialIDMetadataKey, value: credentialID),
            BinaryMetadataEntry(key: credentialSecretMetadataKey, value: secret),
        ]
    }

    public var description: String {
        "ServiceCredential(credential_id_len=\(credentialID.count), secret=[REDACTED])"
    }
}

File: sdk/kotlin/src/main/kotlin/org/ucr/sdk/ServiceCredential.kt
package org.ucr.sdk

const val CREDENTIAL_ID_METADATA_KEY: String = "ucr-service-credential-id-bin"
const val CREDENTIAL_SECRET_METADATA_KEY: String = "ucr-service-credential-secret-bin"

data class BinaryMetadataEntry(val key: String, val value: ByteArray)

/** Opaque Service Principal credential for generated ucr.v1 clients. */
class ServiceCredential(credentialId: ByteArray, secret: ByteArray) {
    private val credentialId = credentialId.copyOf()
    private val secret = secret.copyOf()

    fun binaryMetadata(): List<BinaryMetadataEntry> = listOf(
        BinaryMetadataEntry(CREDENTIAL_ID_METADATA_KEY, credentialId.copyOf()),
        BinaryMetadataEntry(CREDENTIAL_SECRET_METADATA_KEY, secret.copyOf()),
    )

    override fun toString(): String =
        "ServiceCredential(credential_id_len=${credentialId.size}, secret=[REDACTED])"
}
